File: Player/Android/Quber/Quber_Player/app/src/main/java/kr/co/turtlelab/andowsignage/tools/SecureJsonTools.java
package kr.co.turtlelab.andowsignage.tools;

import android.util.Base64;

import com.google.gson.Gson;

import java.io.File;
import java.io.FileInputStream;
import java.io.ByteArrayOutputStream;
import java.nio.charset.Charset;
import java.security.MessageDigest;
import java.util.Arrays;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public final class SecureJsonTools {
    private static final int IV_SIZE = 16;
    private static final int HMAC_SIZE = 32;
    private static final String DEFAULT_PASSPHRASE = "ninja04!9akftp!";
    private static final byte[] FAST_MAGIC = new byte[]{'N', 'H', 'Y', '2'};
    private static final Charset UTF8 = Charset.forName("UTF-8");
    private static final Gson GSON = new Gson();

    private SecureJsonTools() {
    }

    public static <T> T readEncryptedJson(File file, Class<T> type) throws Exception {
        if (file == null || !file.exists() || type == null) {
            return null;
        }

        String base64 = new String(readAllBytes(file), UTF8).trim();
        if (base64.length() < 1) {
            return null;
        }

        byte[] encrypted = Base64.decode(base64, Base64.DEFAULT);
        String json = decrypt(encrypted);
        if (json == null || json.length() < 1) {
            return null;
        }

        return GSON.fromJson(json, type);
    }

    private static String decrypt(byte[] encryptedPayload) throws Exception {
        if (!isFastPayload(encryptedPayload)) {
            return null;
        }

        return decryptFast(encryptedPayload);
    }

    private static String decryptFast(byte[] encryptedPayload) throws Exception {
        if (encryptedPayload == null || encryptedPayload.length < FAST_MAGIC.length + IV_SIZE + HMAC_SIZE) {
            return null;
        }

        byte[] encKey = computeSha256(DEFAULT_PASSPHRASE + ":enc");
        byte[] macKey = computeSha256(DEFAULT_PASSPHRASE + ":mac");

        int cipherOffset = FAST_MAGIC.length + IV_SIZE;
        int cipherLength = encryptedPayload.length - cipherOffset - HMAC_SIZE;
        if (cipherLength <= 0) {
            return null;
        }

        byte[] providedMac = Arrays.copyOfRange(encryptedPayload, cipherOffset + cipherLength, encryptedPayload.length);
        byte[] expectedMac = hmacSha256(macKey, encryptedPayload, 0, cipherOffset + cipherLength);
        if (!constantTimeEquals(providedMac, expectedMac)) {
            return null;
        }

        byte[] iv = Arrays.copyOfRange(encryptedPayload, FAST_MAGIC.length, FAST_MAGIC.length + IV_SIZE);
        byte[] cipherBytes = Arrays.copyOfRange(encryptedPayload, cipherOffset, cipherOffset + cipherLength);

        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(encKey, "AES"), new IvParameterSpec(iv));
        byte[] plain = cipher.doFinal(cipherBytes);
        return new String(plain, UTF8);
    }

    private static boolean isFastPayload(byte[] payload) {
        if (payload == null || payload.length < FAST_MAGIC.length) {
            return false;
        }
        for (int i = 0; i < FAST_MAGIC.length; i++) {
            if (payload[i] != FAST_MAGIC[i]) {
                return false;
            }
        }
        return true;
    }

    private static byte[] hmacSha256(byte[] key, byte[] data, int offset, int count) throws Exception {
        Mac hmac = Mac.getInstance("HmacSHA256");
        hmac.init(new SecretKeySpec(key, "HmacSHA256"));
        hmac.update(data, offset, count);
        return hmac.doFinal();
    }

    private static byte[] computeSha256(String value) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        return digest.digest((value == null ? "" : value).getBytes(UTF8));
    }

    private static boolean constantTimeEquals(byte[] left, byte[] right) {
        if (left == null || right == null || left.length != right.length) {
            return false;
        }

        int diff = 0;
        for (int i = 0; i < left.length; i++) {
            diff |= left[i] ^ right[i];
        }

        return diff == 0;
    }

    private static byte[] readAllBytes(File file) throws Exception {
        FileInputStream input = null;
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        try {
            input = new FileInputStream(file);
            byte[] buffer = new byte[8192];
            int read;
            while ((read = input.read(buffer)) > 0) {
                output.write(buffer, 0, read);
            }
            return output.toByteArray();
        } finally {
            if (input != null) {
                input.close();
            }
        }
    }
}

File: Player/Windows/NewHyOn_Player/TurtleTools/SecureJsonTools.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using Newtonsoft.Json;

namespace TurtleTools
{
    public static class SecureJsonTools
    {
        private const int IvSize = 16;
        private const int HmacSize = 32;
        private const string DefaultPassphrase = "ninja04!9akftp!";
        private static readonly byte[] FastMagic = Encoding.ASCII.GetBytes("NHY2");

        public static T ReadEncryptedJson<T>(string filePath, string passphrase = null)
        {
            if (string.IsNullOrWhiteSpace(filePath) || File.Exists(filePath) == false)
            {
                return default(T);
            }

            string payloadBase64 = File.ReadAllText(filePath, Encoding.UTF8);
            byte[] encrypted = Convert.FromBase64String(payloadBase64);
            string json = Decrypt(encrypted, passphrase);

            if (json == null)
            {
                return default(T);
            }

            return JsonConvert.DeserializeObject<T>(json);
        }

        private static string Decrypt(byte[] encryptedPayload, string passphrase)
        {
            if (IsFastPayload(encryptedPayload) == false)
            {
                return null;
            }

            return DecryptFast(encryptedPayload, passphrase);
        }

        private static string DecryptFast(byte[] encryptedPayload, string passphrase)
        {
            if (encryptedPayload == null || encryptedPayload.Length < FastMagic.Length + IvSize + HmacSize)
            {
                return null;
            }

            DeriveFastKeys(passphrase, out byte[] encKey, out byte[] macKey);

            int cipherOffset = FastMagic.Length + IvSize;
            int cipherLength = encryptedPayload.Length - cipherOffset - HmacSize;
            if (cipherLength <= 0)
            {
                return null;
            }

            byte[] providedMac = new byte[HmacSize];
            Buffer.BlockCopy(encryptedPayload, cipherOffset + cipherLength, providedMac, 0, HmacSize);
            byte[] expectedMac = ComputeHmac(macKey, encryptedPayload, 0, cipherOffset + cipherLength);
            if (!ConstantTimeEquals(providedMac, expectedMac))
            {
                return null;
            }

            byte[] iv = new byte[IvSize];
            Buffer.BlockCopy(encryptedPayload, FastMagic.Length, iv, 0, IvSize);
            byte[] cipherBytes = new byte[cipherLength];
            Buffer.BlockCopy(encryptedPayload, cipherOffset, cipherBytes, 0, cipherLength);

            using (Aes aes = Aes.Create())
            {
                aes.KeySize = 256;
                aes.Mode = CipherMode.CBC;
                aes.Padding = PaddingMode.PKCS7;
                aes.Key = encKey;
                aes.IV = iv;

                using (ICryptoTransform decryptor = aes.CreateDecryptor())
                {
                    byte[] plainBytes = decryptor.TransformFinalBlock(cipherBytes, 0, cipherBytes.Length);
                    return Encoding.UTF8.GetString(plainBytes);
                }
            }
        }

        private static bool IsFastPayload(byte[] payload)
        {
            if (payload == null || payload.Length < FastMagic.Length)
            {
                return false;
            }

            for (int i = 0; i < FastMagic.Length; i++)
            {
                if (payload[i] != FastMagic[i])
                {
                    return false;
                }
            }
            return true;
        }

        private static void DeriveFastKeys(string passphrase, out byte[] encKey, out byte[] macKey)
        {
            string secret = string.IsNullOrEmpty(passphrase) ? DefaultPassphrase : passphrase;
            encKey = ComputeSha256(secret + ":enc");
            macKey = ComputeSha256(secret + ":mac");
        }

        private static byte[] ComputeSha256(string value)
        {
            using (SHA256 sha = SHA256.Create())
            {
                return sha.ComputeHash(Encoding.UTF8.GetBytes(value ?? string.Empty));
            }
        }

        private static byte[] ComputeHmac(byte[] key, byte[] data, int offset, int count)
        {
            using (HMACSHA256 hmac = new HMACSHA256(key))
            {
                return hmac.ComputeHash(data, offset, count);
            }
        }

        private static bool ConstantTimeEquals(byte[] left, byte[] right)
        {
            if (left == null || right == null || left.Length != right.Length)
            {
                return false;
            }

            int diff = 0;
            for (int i = 0; i < left.Length; i++)
            {
                diff |= left[i] ^ right[i];
            }

            return diff == 0;
        }
    }
}
